src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpUtilities.cs

```diff
@@ -83,54 +84,63 @@ private static ulong GetMaskAsLong(ReadOnlySpan<byte> bytes)
     {
         if (span.IsEmpty)
         {
             return string.Empty;
         }
 
-        fixed (byte* source = &MemoryMarshal.GetReference(span))
+        var str = string.Create(span.Length, span, (destination, spanPtr) =>
         {
-            return string.Create(span.Length, new IntPtr(source), s_getHeaderName);
-        }
-    }
-
-    private static unsafe void GetHeaderName(Span<char> buffer, IntPtr state)
-    {
-        fixed (char* output = &MemoryMarshal.GetReference(buffer))
-        {
-            // This version of AsciiUtilities returns null if there are any null (0 byte) characters
-            // in the string
-            if (!StringUtilities.TryGetAsciiString((byte*)state.ToPointer(), output, buffer.Length))
+            if (Ascii.ToUtf16(spanPtr, destination, out _) != OperationStatus.Done)
             {
                 KestrelBadHttpRequestException.Throw(RequestRejectionReason.InvalidCharactersInHeaderName);
             }
+        });
+
+        if (str.Contains('\0'))
+        {
+            KestrelBadHttpRequestException.Throw(RequestRejectionReason.InvalidCharactersInHeaderName);
         }
+        return str;
     }
 
-    public static string GetAsciiStringNonNullCharacters(this Span<byte> span)
-        => StringUtilities.GetAsciiStringNonNullCharacters(span);
+    // Null checks must be done independently of this method (if required)
+    public static string GetAsciiString(this Span<byte> span)
+        => StringUtilities.GetAsciiString(span);
 
-    public static string GetAsciiOrUTF8StringNonNullCharacters(this ReadOnlySpan<byte> span)
-        => StringUtilities.GetAsciiOrUTF8StringNonNullCharacters(span, DefaultRequestHeaderEncoding);
+    public static string GetAsciiOrUTF8String(this ReadOnlySpan<byte> span)
+        => StringUtilities.GetAsciiOrUTF8String(span, DefaultRequestHeaderEncoding);
 
     public static string GetRequestHeaderString(this ReadOnlySpan<byte> span, string name, Func<string, Encoding?> encodingSelector, bool checkForNewlineChars)
     {
         string result;
         if (ReferenceEquals(KestrelServerOptions.DefaultHeaderEncodingSelector, encodingSelector))
         {
-            result = span.GetAsciiOrUTF8StringNonNullCharacters(DefaultRequestHeaderEncoding);
+            result = span.GetAsciiOrUTF8String(DefaultRequestHeaderEncoding);
         }
         else
         {
             result = span.GetRequestHeaderStringWithoutDefaultEncodingCore(name, encodingSelector);
         }
 
         // New Line characters (CR, LF) are considered invalid at this point.
-        if (checkForNewlineChars && ((ReadOnlySpan<char>)result).IndexOfAny('\r', '\n') >= 0)
+        // Null characters are also not allowed.
+        var invalidCharIndex = checkForNewlineChars ?
+            ((ReadOnlySpan<char>)result).IndexOfAny(_nullAndNewLineSearchValues)
+            : ((ReadOnlySpan<char>)result).IndexOf('\0');
+
+        if (invalidCharIndex >= 0)
```

Comment:
Assuming that `if (invalidCharIndex >= 0)` is the rare-path: maybe split the code in the if-block to a separate non-inlineable method, so that this method gets potentially inlineable?

Reply:
Can we use the ThrowHelper pattern for just the exceptions to achieve the same effect?